File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/08-Stack/LeetCode/LeetCode-0155.py
class MinStack:
    
    def __init__(self):
        self.S1 = []
        self.S2 = []  # Monotonic stack    

    def push(self, val: int) -> None:
        self.S1.append(val)
        if len(self.S2) == 0 or val <= self.S2[-1]:
            self.S2.append(val)        

    def pop(self) -> None:
        x = self.S1.pop()
        if x == self.S2[-1]:
            self.S2.pop()        

    def top(self) -> int:
        return self.S1[-1]        

    def getMin(self) -> int:
        return self.S2[-1]                


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/08-Stack/LeetCode/LeetCode-0155.java
class MinStack {
    Stack<Integer> S1 = new Stack<>();
    Stack<Integer> S2 = new Stack<>();  // Monotonic stack

    public MinStack() {}
    
    public void push(int val) {
        S1.push(val);
        if (S2.isEmpty() || val <= S2.peek()) S2.push(val);        
    }
    
    public void pop() {
        int x = S1.pop();
        if (x == S2.peek()) S2.pop();
    }
    
    public int top() {
        return S1.peek();        
    }
    
    public int getMin() {
        return S2.peek();        
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */